### PR TITLE
(feat)  O3-3113: Add disallowDecimals validator

### PR DESCRIPTION
**Pull request by jabahum**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR implements a validator in the form engine that validates whether decimal values should be allowed in numeric obs fields. This resembles the Angular form engine's similarly named [disallowDecimals validator](https://github.com/openmrs/openmrs-ngx-formentry/pull/135).

The Carbon [NumberInput](https://angular.carbondesignsystem.com/?path=/docs/components-number-input--docs) doesn't provide any special handling for decimal values, so this validator can step in if you want to mandate that a field should not support decimals, for example. A scenario where you'd want to use this is when representing a numeric field that's backed by a concept that has `allowDecimal` set to false, such as Platelets below:

platelets

```
{
  "name": "Test Form 1",
  "description": "Test Form 1 Description",
  "version": "1",
  "published": true,
  "retired": false,
  "encounter": "Consultation",
  "pages": [
    {
      "label": "Page 1",
      "sections": [
        {
          "label": "Section 1",
          "isExpanded": "true",
          "questions": [
            {
              "label": "CD4 count :",
              "type": "obs",
              "id": "cd4count",
              "questionOptions": {
                "concept": "161011AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
                "rendering": "number",
                "disallowDecimals": "true"
              },
              "validators": []
            }
          ]
        }
      ]
    }
  ],
  "processor": "EncounterFormProcessor",
  "referencedForms": [],
  "encounterType": "67a71486-1a54-468f-ac3e-7091a9a79584"
}
```

### CODE REVIEW

File: src/validators/form-validator.ts [hunk 1]
```diff
@@ -51,6 +51,18 @@ export function numberInputRangeValidator(min: number, max: number, inputValue:
     ];
   }
 
+  if (field.questionOptions.disallowDecimals) {
+    if (!isEmpty(inputValue) && inputValue.toString().includes('.')) {
+      return [
+        {
+          resultType: 'error',
+          errCode: fieldOutOfBoundErrCode,
+          message: `Decimal values are not allowed`,
```
Comment (pirupius) [inline src/validators/form-validator.ts:60]: Add translations 😈

**Comment by denniskigen** [comment]
@jabahum I've updated the PR description with extra context.

**Comment by jabahum** [comment]
> @jabahum I've updated the PR description with extra context. It'd be cool if you could add a sample schema that shows how to leverage this validator, as well as a short video of it in action.

This has been added @denniskigen

File: src/validators/form-validator.ts [hunk 2]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
+export function numberInputRangeValidator(min: number, max: number, inputValue: number, field?: FormField) {
   if (!Number.isNaN(min) && inputValue < min) {
     return [
       {
-        resultType: 'error',
+        // t('error', 'error')
+        resultType: translateFrom(moduleName, 'error', 'error'),
         errCode: fieldOutOfBoundErrCode,
-        message: `Value must be greater than ${min}`,
+        // t('minValue', 'Value must be lower than {{min}}', {min})
```
Comment (pirupius) [inline src/validators/form-validator.ts:42]: This string seems to be reversed. I believe you meant higher than min

File: src/validators/form-validator.ts [hunk 3]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
   }
 
+  if (field.questionOptions.disallowDecimals) {
+    if (!isEmpty(inputValue) && inputValue.toString().includes('.')) {
+      return [
+        {
+          // t('error', 'error')
+          resultType: translateFrom(moduleName, 'error', 'error'),
+          errCode: fieldOutOfBoundErrCode,
+          // t('decimalValue', 'Decimal values are not allowed ${inputValue}', {inputValue})
```
Comment (pirupius) [inline src/validators/form-validator.ts:71]:
```suggestion
          // t('decimalValue', 'Decimal values are not allowed.',)
```

File: src/validators/form-validator.ts [hunk 4]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
   if (!Number.isNaN(min) && inputValue < min) {
     return [
       {
-        resultType: 'error',
+        // t('error', 'error')
+        resultType: translateFrom(moduleName, 'error', 'error'),
         errCode: fieldOutOfBoundErrCode,
-        message: `Value must be greater than ${min}`,
+        // t('maxValue', 'Value must be lower than {{max}}', {max})
+        message: translateFrom(moduleName, 'maxValue', 'Value must be greater than {{max}}', {
```
Comment (pirupius) [inline src/validators/form-validator.ts:43]:
```suggestion
        // t('maxValue', 'Value must be lower than {{max}}', {max})
        message: translateFrom(moduleName, 'maxValue', 'Value must be lower than {{max}}', {
```

File: src/validators/form-validator.ts [hunk 5]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
   if (!Number.isNaN(max) && inputValue > max) {
     return [
       {
-        resultType: 'error',
+        // t('error', 'error')
+        resultType: translateFrom(moduleName, 'error', 'error'),
         errCode: fieldOutOfBoundErrCode,
-        message: `Value must be lower than ${max}`,
+        // t('minValue', 'Value must be lower than ${min}', {min})
+        message: translateFrom(moduleName, 'minValue', 'Value must be lower than ${min}', {
```
Comment (pirupius) [inline src/validators/form-validator.ts:57]:
```suggestion
        // t('minValue', 'Value must be greater than ${min}', {min})
        message: translateFrom(moduleName, 'minValue', 'Value must be greater than ${min}', {
```

File: src/validators/form-validator.ts [hunk 6]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
 
+  if (field.questionOptions.disallowDecimals) {
+    if (!isEmpty(inputValue) && inputValue.toString().includes('.')) {
+      return [
+        {
+          // t('error', 'error')
+          resultType: translateFrom(moduleName, 'error', 'error'),
+          errCode: fieldOutOfBoundErrCode,
+          // t('decimalValue', 'Decimal values are not allowed ')
+          message: translateFrom(moduleName, 'decimalValue', 'Decimal values are not allowed', {
```
Comment (pirupius) [inline src/validators/form-validator.ts:72]:
```suggestion
          // t('decimalValueNotAllowed', 'Decimal values are not allowed for this field')
          message: translateFrom(moduleName, 'decimalValue', 'Decimal values are not allowed for this field', {
```

**Review by pirupius — DISMISSED** [review]

File: src/validators/form-validator.ts [hunk 7]
```diff
@@ -24,33 +26,57 @@ export const FieldValidator: FormFieldValidator = {
...
-        message: `Value must be lower than ${max}`,
+        // t('minValue', 'Value must be lower than ${min}', {min})
+        message: translateFrom(moduleName, 'minValue', 'Value must be lower than ${min}', {
+          max,
+        }),
       },
     ];
   }
 
+  if (field.questionOptions.disallowDecimals) {
```
Comment (samuelmale) [inline src/validators/form-validator.ts:64]: What happens if a user forgets to update this value in the form? In such scenarios, we should infer this value from the concept.

Comment (jabahum) [reply src/validators/form-validator.ts:64]: noted

File: package.json [hunk 8]
```diff
@@ -14,7 +14,7 @@
     "coverage": "yarn test --coverage",
     "analyze": "webpack --mode=production --env.analyze=true",
     "prepare": "husky install",
-    "extract-translations": "i18next 'src/**/*.component.tsx' --config './tools/i18next-parser.config.js'"
+    "extract-translations": "i18next 'src/**/*.{ts,tsx}' --config './tools/i18next-parser.config.js'"
```
Comment (samuelmale) [inline package.json:17]: Did you intend to make this change?

File: src/validators/form-validator.ts [hunk 9]
```diff
@@ -24,33 +27,55 @@ export const FieldValidator: FormFieldValidator = {
...
 };
 
-export function numberInputRangeValidator(min: number, max: number, inputValue: number) {
+export function numberInputRangeValidator(min: number, max: number, inputValue: number, field?: FormField) {
   if (!Number.isNaN(min) && inputValue < min) {
     return [
       {
-        resultType: 'error',
+        // t('error', 'error')
+        resultType: translateFrom(moduleName, 'error', 'error'),
```
Comment (samuelmale) [inline src/validators/form-validator.ts:41]: I'm a bit curious of whether `translateFrom(moduleName, 'error', 'error')` would work here. I think the `moduleName` should point to an app that is consuming this library ie. the "@openmrs/esm-form-engine-app" or if the translations exist in core. cc: @vasharma05 @ibacher

File: package.json (hunk 8 above)
Comment (jabahum) [reply package.json:17]: yes this was intended

Comment (samuelmale) [reply package.json:17]: You don't need this since the translations will be done by the form-engine-app

**Comment by denniskigen** [comment]
@jabahum I've tried filling in the Laboratory Test Results form and I don't see the validation kick in when entering decimal values for the `Platelets` and `MCV` fields, which don't allow decimal values.
 
CleanShot 2024-05-08 at 11  57 50@2x

File: src/validators/form-validator.ts [hunk 10]
```diff
@@ -24,33 +26,53 @@ export const FieldValidator: FormFieldValidator = {
...
   },
 };
 
-export function numberInputRangeValidator(min: number, max: number, inputValue: number) {
+export function numberInputRangeValidator(min: number, max: number, inputValue: number, field?: FormField) {
   if (!Number.isNaN(min) && inputValue < min) {
     return [
       {
-        resultType: 'error',
+        resultType: translateFrom('@openmrs/esm-form-engine-app', 'error', 'error'),
```
Comment (samuelmale) [inline src/validators/form-validator.ts:39]: @jabahum why would you translate this code?

File: src/validators/form-validator.ts [hunk 11]
```diff
@@ -24,33 +26,53 @@ export const FieldValidator: FormFieldValidator = {
...
+        }),
       },
     ];
   }
 
   if (!Number.isNaN(max) && inputValue > max) {
     return [
       {
-        resultType: 'error',
+        resultType: translateFrom('@openmrs/esm-form-engine-app', 'error', 'error'),
```
Comment (samuelmale) [inline src/validators/form-validator.ts:51]: Ditto

File: src/validators/form-validator.ts [hunk 12]
```diff
@@ -93,9 +115,9 @@ export function textInputLengthValidator(minLength: string, maxLength: string, i
 export function addError(errorCode: string, message: string): [{}] {
   return [
     {
-      resultType: 'error',
+      resultType: translateFrom('@openmrs/esm-form-engine-app', 'error', 'error'),
```
Comment (samuelmale) [inline src/validators/form-validator.ts:118]: Ditto

**Comment by denniskigen** [comment]
@jabahum is the Lab Test Results form now getting validated correctly?

**Comment by hadijahkyampeire** [comment]
@denniskigen your issue has been fixed, it was because the question was lacking min and max, so I have set a default since all the validation bases on the question having min and max
Screenshot 2024-05-14 at 21 19 48

**Comment by denniskigen** [comment]
@hadijahkyampeire the validation should be able to work independent of whether min and max values are set.

File: translations/en.json [hunk 13]
```diff
@@ -12,23 +12,30 @@
...
   "loading": "Loading",
+  "maxValue": "Value must be greater than ${max}",
+  "minValue": "Value must be lower than {{min}}",
   "notification": "Notification",
   "ordersSaved": "Order(s) saved sucessfully",
   "patientIdentifiersSaved": "Patient identifier(s) saved sucessfully",
   "preview": "Preview",
   "previousValue": "Previous value:",
-  "remove": "Remove",
+  "removeGroup": "Remove group",
```
Comment (samuelmale) [inline translations/en.json:38]: This was a rebase residue. Can you revert to "remove"?

File: src/validators/form-validator.ts [hunk 14]
```diff
@@ -6,7 +6,7 @@ export const fieldOutOfBoundErrCode = 'field.outOfBound';
 
 export const FieldValidator: FormFieldValidator = {
   validate: (field: FormField, value: any) => {
-    if (field.meta.submission?.unspecified) {
+    if (field?.meta?.submission?.unspecified) {
```
Comment (samuelmale) [inline src/validators/form-validator.ts:9]: Can `meta` be null at this point?

File: translations/en.json [hunk 15]
```diff
@@ -12,23 +12,30 @@
   "closesNotification": "Closes notification",
   "createdRecord": "Record created",
   "createdRecordDescription": "A new encounter was created",
+  "decimalValue": "Decimal values are not allowed ${inputValue}",
+  "error": "error",
+  "errorDescription": "",
```
Comment (samuelmale) [inline translations/en.json:17]: Can you run the script below to update this file?

```
yarn run extract-translations
```

**Review by samuelmale — APPROVED** [review]
Just a few comments

**Pull request merged**